### ISSUE TITLE
Annotations - Abstract class support?

### ISSUE THREAD

**Issue by manafire** [Stacked-Org/stacked#228]
One of the things that was possible with the old injectable/locator setup was [binding abstract classes to implementations](https://pub.dev/packages/injectable#binding-abstract-classes-to-implementations), like so:

```
@LazySingleton(as: AuthService)
FirebaseAuthService get authService;
```

Wondering if there's any merit to add support for an `as:` param (or `asClassType`)?  .

```
@StackedApp(
  dependencies: [
    LazySingleton(classType: FirebaseAuthService, as: AuthService),
])

// would generate in app.locator.dart:
// -->  locator.registerLazySingleton<AuthService>(() => FirebaseAuthService());
```

I forked Stacked and took a crack at it but just couldn't quite get it.  It's an easy workaround by just adding my own extra customLocatorSetup() function in main.dart with these special lines but obviously, not as cleaner as having it part of the annotation.  Thought I would suggest this, if you think it worthwhile.

**Comment by FilledStacks** [comment]
@manafire good catch. I should have probably included that in the initial setup. There's definitely merit to it. I've not been using the abstractions / interfaces in most cases because of the lack of maintenance points it gives me. But that's definitely something we should add in. I probably won't have time today but I might tomorrow, it's my content day so I'm doing package management and working on the new video. 

If you want to take another crack at it you can open `getit_locator_generator.dart` then in `_getLocatorRegistrationStringForType` you can update the code and add the below before the switch

```dart
final hasAbstratedType = dependencyDefinition.asType != null;
final abstractionType = hasAbstratedType ? dependencyDefinition.abstractedTypeClassName : '';
```
Before that you'll have to serialise the new asType and abstractedTypeClassName  properties on the `dependencyDefinition`.  Once that's all in there you can update the registration string from 

```dart
'locator.registerLazySingleton(() => ${dependencyDefinition.className}());'
```
to
```dart
'locator.registerLazySingleton$abstractionType(() => ${dependencyDefinition.className}());'
```
for all 4 cases. The actual function to be called on the locator should actually be abstracted out as well so we can determine that separately but that refactor can be done in my spare time :D 

Let me know the progress, I'll probably tackle this around (hopefully) 15:00-16:30 SAST tomorrow 25 Feb . If there's no update I'll assume you didn't do it and I'll implement that and push up the code.

**Comment by manafire** [comment]
Thanks, Dane.  I haven't figured out to wire up build_runner to VS Code to debug with breakpoints so I've been flying blind and the code generation / annotations is entirely new to me.  I managed to get it working though and submitted a pull request. https://github.com/FilledStacks/stacked/pull/230 (hope I did that correctly?)

I had to ensure the abstracted service type was imported in the generated file and added an extra `abstractedImport` property to `DependencyConfig` to do so.  I thought about changing `final String import` to `List<String>` perhaps but that seemed to create more problems and complexity than I had the time to deal with (got to get to my day job!).
